### ISSUE TITLE
no-dynamic-module-names

### ISSUE THREAD

**Issue by theengineear** [eslint/eslint#19051]
### Rule details

Disallow passing variables into calls to `import()`.

### Related ECMAScript feature

Dynamic import / `import()`.

### What type of rule is this?

Warns about a potential problem

### Example code

```js
class Modularized {
  static personDetailsPath = './person/details.js'; // Oops, should be `../person/detail.js`!

  static async getPersonDetails(personId) {
    // This would warn because it makes it harder to debug relative import paths.
    const module = await import(Modularized.personDetailsPath);
    return await module.getDetails(personId);
  }

  static async getEmployerDetails(employerId) {
    // This is easier to debug — e.g., IDEs can easily tell you if this path exists.
    const module = await import('../employer.details');
    return await module.getDetails(employerId);
  }
}
```

### Why should this rule be in the core instead of a plugin?

Static import declarations cannot be given a variable module name. This restriction improves static analysis and developer ergonomics. It should be possible (recommended?) to analogously restrict the functional equivalent (dynamic import / `import()`) to reap similar benefits.

### Participation

- [x] I am willing to submit a pull request to implement this rule.

### Additional comments

Thank you very much for your consideration here! I checked the box as “willing to implement” — but have never contributed to this codebase before. My false sense of confidence comes from my rationale that (1) `import` is not a valid variable name and (2) this rule would just need to check that a simple string is passed as the first argument. It would be quite similar to the [no-dynamic-require](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-dynamic-require.md) plugin rule.

**Comment by mdjermanovic** [comment]
Hi @theengineear, thanks for the issue!

> Static import declarations cannot be given a variable module name. This restriction improves static analysis and developer ergonomics. It should be possible (recommended?) to analogously restrict the functional equivalent (dynamic import / `import()`) to reap similar benefits.

I think this depends on the use case. One of the use cases described in the [import() proposal](https://github.com/tc39/proposal-dynamic-import?tab=readme-ov-file) is "factors only known at runtime (such as the user's language)" in which case restricting dynamic imports to predefined strings may not be possible. One example of this use case is ESLint itself - config files are loaded using `import()`, but the location of the config file is known only at runtime.

https://github.com/eslint/eslint/blob/ee0a77ea3caa5838bab704b54a577eefbed58f68/lib/config/config-loader.js#L202

So I think this rule wouldn't be generally applicable, and as such is not suitable for the core. You can already enforce this restriction by configuring the [`no-restricted-syntax`](https://eslint.org/docs/latest/rules/no-restricted-syntax) rule with `ImportExpression > :not(Literal).source` selector: [Playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLCB7XG4gICAgc2VsZWN0b3I6IFwiSW1wb3J0RXhwcmVzc2lvbiA+IDpub3QoTGl0ZXJhbCkuc291cmNlXCIsXG4gICAgbWVzc2FnZTogXCJVbmV4cGVjdGVkIHZhcmlhYmxlIG1vZHVsZSBuYW1lLlwiXG59XSAqL1xuXG5jbGFzcyBNb2R1bGFyaXplZCB7XG4gIHN0YXRpYyBwZXJzb25EZXRhaWxzUGF0aCA9ICcuL3BlcnNvbi9kZXRhaWxzLmpzJzsgLy8gT29wcywgc2hvdWxkIGJlIGAuLi9wZXJzb24vZGV0YWlsLmpzYCFcblxuICBzdGF0aWMgYXN5bmMgZ2V0UGVyc29uRGV0YWlscyhwZXJzb25JZCkge1xuICAgIC8vIFRoaXMgd291bGQgd2FybiBiZWNhdXNlIGl0IG1ha2VzIGl0IGhhcmRlciB0byBkZWJ1ZyByZWxhdGl2ZSBpbXBvcnQgcGF0aHMuXG4gICAgY29uc3QgbW9kdWxlID0gYXdhaXQgaW1wb3J0KE1vZHVsYXJpemVkLnBlcnNvbkRldGFpbHNQYXRoKTtcbiAgICByZXR1cm4gYXdhaXQgbW9kdWxlLmdldERldGFpbHMocGVyc29uSWQpO1xuICB9XG5cbiAgc3RhdGljIGFzeW5jIGdldEVtcGxveWVyRGV0YWlscyhlbXBsb3llcklkKSB7XG4gICAgLy8gVGhpcyBpcyBlYXNpZXIgdG8gZGVidWcg4oCUIGUuZy4sIElERXMgY2FuIGVhc2lseSB0ZWxsIHlvdSBpZiB0aGlzIHBhdGggZXhpc3RzLlxuICAgIGNvbnN0IG1vZHVsZSA9IGF3YWl0IGltcG9ydCgnLi4vZW1wbG95ZXIuZGV0YWlscycpO1xuICAgIHJldHVybiBhd2FpdCBtb2R1bGUuZ2V0RGV0YWlscyhlbXBsb3llcklkKTtcbiAgfVxufSIsIm9wdGlvbnMiOnsicnVsZXMiOnt9LCJsYW5ndWFnZU9wdGlvbnMiOnsic291cmNlVHlwZSI6InNjcmlwdCIsInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX19fX0=).

@eslint/eslint-team thoughts?

**Comment by nzakas** [comment]
We decided a while back that we weren't going to be adding new core rules that simply disallowed syntax because 1) `no-restricted-syntax` is available and 2) creating custom rules to disallow syntax is very easy. So, those are the two options for getting the rule that you want.

**Comment by theengineear** [comment]
Thanks so much for the quick response — I’m in agreement with closing this issue. I’m embarrassed that I wasn’t able to find the solution via `no-restricted-syntax` myself 😳 

Sorry for the distraction!
